sys/include/auto_init_macros.h

```diff
@@ -0,0 +1,432 @@
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_SHT1X)
+#define AUTO_INIT_PRIO_SAUL \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_SAUL)
+#define AUTO_INIT_PRIO_GNRC_RPL \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_GNRC_RPL)
+#define AUTO_INIT_PRIO_CAN \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_CAN)
+#define AUTO_INIT_PRIO_SUIT \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_SUIT)
+#define AUTO_INIT_PRIO_SECURITY \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_SECURITY)
+#define AUTO_INIT_PRIO_TEST_UTILS_INTERACTIVE_SYNC \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_TEST_UTILS_INTERACTIVE_SYNC)
+#define AUTO_INIT_PRIO_DHCPV6_CLIENT \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_DHCPV6_CLIENT)
+#define AUTO_INIT_PRIO_DHCPV6_RELAY \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_DHCPV6_RELAY)
+#define AUTO_INIT_PRIO_GNRC_DHCPV6_CLIENT_SIMPLE_PD \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_GNRC_DHCPV6_CLIENT_SIMPLE_PD)
+#define AUTO_INIT_PRIO_GNRC_IPV6_AUTO_SUBNETS \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_GNRC_IPV6_AUTO_SUBNETS)
+#define AUTO_INIT_PRIO_MULTIMEDIA \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_MULTIMEDIA)
+#define AUTO_INIT_PRIO_SCREEN \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_SCREEN)
+#define AUTO_INIT_PRIO_BENCHMARK_UDP \
+    AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_HIGHEST, CONFIG_AUTO_INIT_PRIO_BENCHMARK_UDP)
+/** @} */
+
+#if ENABLE_DEBUG || defined(DOXYGEN)
+/**
+ * @brief   Add a module to the auto-initialization array
+ *
+ * @param   module      Module name
+ * @param   prio        Priority level
+ * @param   function    Function to be called on initialization
+ *
+ * @experimental
+ */
+#define AUTO_INIT_MODULE(module, prio, function)                                        \
+    static inline void CONCAT(CONCAT(_debug_, prio), function)(void) {                  \
+        DEBUG("auto_init: %s\n", __func__);                                             \
+        function();                                                                     \
+    }                                                                                   \
+    XFA_CONST(auto_init_xfa, prio)                                                      \
+    auto_init_fn CONCAT(CONCAT(auto_init_xfa, prio), module) =                          \
+        (auto_init_fn)CONCAT(CONCAT(_debug_, prio), function)
+#else
+#define AUTO_INIT_MODULE(module, prio, function)                                        \
+    XFA_CONST(auto_init_xfa, prio)                                                      \
+    auto_init_fn CONCAT(CONCAT(auto_init_xfa, prio), module) = (auto_init_fn)function
+#endif
```

Comment:
```suggestion
/**
 * @brief   Add a module to the auto-initialization array
 *
 * @param   module      Module name
 * @param   prio        Priority level
 * @param   function    Function to be called on initialization
 *
 * @experimental
 */
#if AUTOINIT_DEBUG || defined(DOXYGEN)
#define AUTO_INIT_MODULE(module, prio, function)                                        \
    AUTO_INIT_MODULE_DEBUG(module, prio, function)
#else
#define AUTO_INIT_MODULE(module, prio, function)                                        \
    XFA_CONST(auto_init_xfa, prio)                                                      \
    auto_init_fn CONCAT(CONCAT(auto_init_xfa, prio), module) = (auto_init_fn)function
#endif
    
#define AUTO_INIT_MODULE_DEBUG(module, prio, function)                                        \
    static inline void CONCAT(CONCAT(_debug_, prio), function)(void) {                  \
        DEBUG_PUTS("auto_init: " module);                                             \
        function();                                                                     \
    }                                                                                   \
    XFA_CONST(auto_init_xfa, prio)                                                      \
    auto_init_fn CONCAT(CONCAT(auto_init_xfa, prio), module) =                          \
        (auto_init_fn)CONCAT(CONCAT(_debug_, prio), function)
```

Reply:
Ok, and `AUTOINIT_DEBUG` should be defined with `CFLAGS`?
Using `DEBUG_PUTS()` we still need `ENABLE_DEBUG` to be `!= 0`.
Should I make it a simple `puts()` instead. Then we do not need `debug.h`.

Reply:
Or have `CONFIG_AUTO_INIT_DEBUG` and expose it to Kconfig.